Sources/Testing/Running/Runner.Context.swift

```diff
@@ -0,0 +1,226 @@
+      let id = all.nextID
+      all.nextID += 1
+      all.instances[id] = self
+      return id
+    }
+  }
+
+  /// Remove this instance from ``Configuration/all``.
+  ///
+  /// - Parameters:
+  ///   - id: The unique identifier of this instance, as previously returned by
+  ///     `_addToAll()`. If `nil`, this function has no effect.
+  private func _removeFromAll(identifiedBy id: UInt64?) {
+    if let id {
+      Self.$_all.withLock { all in
+        _ = all.instances.removeValue(forKey: id)
+      }
+    }
+  }
+}
+
+// MARK: - Current test and test case
+
+extension Test {
+  /// The test that is running on the current task, if any.
+  public static var current: Self? {
+    Runner.Context.current.test
+  }
+
+  /// Call a function while the value of ``Test/current`` is set.
+  ///
+  /// - Parameters:
+  ///   - test: The new value to set for ``Test/current``.
+  ///   - body: A function to call.
+  ///
+  /// - Returns: Whatever is returned by `body`.
+  ///
+  /// - Throws: Whatever is thrown by `body`.
+  static func withCurrent<R>(_ test: Self, perform body: () throws -> R) rethrows -> R {
+    var context = Runner.Context.current
+    context.test = test
+    return try Runner.Context.$current.withValue(context, operation: body)
+  }
+
+  /// Call a function while the value of ``Test/current`` is set.
+  ///
+  /// - Parameters:
+  ///   - test: The new value to set for ``Test/current``.
+  ///   - body: A function to call.
+  ///
+  /// - Returns: Whatever is returned by `body`.
+  ///
+  /// - Throws: Whatever is thrown by `body`.
+  static func withCurrent<R>(_ test: Self, perform body: () async throws -> R) async rethrows -> R {
+    var context = Runner.Context.current
+    context.test = test
+    return try await Runner.Context.$current.withValue(context, operation: body)
+  }
+}
+
+extension Test.Case {
+  /// The test case that is running on the current task, if any.
+  public static var current: Self? {
+    Runner.Context.current.testCase
+  }
+
+  /// Call a function while the value of ``Test/Case/current`` is set.
+  ///
+  /// - Parameters:
+  ///   - testCase: The new value to set for ``Test/Case/current``.
+  ///   - body: A function to call.
+  ///
+  /// - Returns: Whatever is returned by `body`.
+  ///
+  /// - Throws: Whatever is thrown by `body`.
+  static func withCurrent<R>(_ testCase: Self, perform body: () throws -> R) rethrows -> R {
+    var context = Runner.Context.current
+    context.testCase = testCase
+    return try Runner.Context.$current.withValue(context, operation: body)
+  }
+
+  /// Call a function while the value of ``Test/Case/current`` is set.
+  ///
+  /// - Parameters:
+  ///   - testCase: The new value to set for ``Test/Case/current``.
+  ///   - body: A function to call.
+  ///
+  /// - Returns: Whatever is returned by `body`.
+  ///
+  /// - Throws: Whatever is thrown by `body`.
+  static func withCurrent<R>(_ testCase: Self, perform body: () async throws -> R) async rethrows -> R {
```

Comment:
It occurs to me we could probably merge these two implementations into one that takes a `WriteableKeyPath<Runner.Context, T>`. I don't know if that's worth exploring (if it is, I'll look at it for a future PR, no need to do it today.)

Reply:
Sounds good, I'll leave that to you to explore